File: scripts/governance/check_circular_deps.py
#!/usr/bin/env python3
"""Dependency Health — detect circular import cycles in Python source"""

import ast
import sys
from pathlib import Path


def build_import_graph(directory: Path) -> dict[str, list[str]]:
    graph = {}
    for filepath in directory.rglob("*.py"):
        if any(p in str(filepath) for p in ["__pycache__", ".venv", "venv"]):
            continue
        try:
            tree = ast.parse(filepath.read_text())
        except SyntaxError:
            continue
        imports = []
        for node in ast.walk(tree):
            if isinstance(node, ast.ImportFrom) and node.module:
                # Convert module path to file path
                parts = node.module.split(".")
                candidate = directory / Path(*parts)
                for ext in [".py", "/__init__.py"]:
                    full = Path(str(candidate) + ext)
                    if full.exists():
                        imports.append(str(full))
                        break
        graph[str(filepath)] = imports
    return graph


def detect_cycles(graph: dict) -> list[list[str]]:
    cycles = []
    visited = set()
    stack = set()

    def dfs(node: str, path: list[str]):
        if node in stack:
            start = path.index(node)
            cycles.append(path[start:] + [node])
            return
        if node in visited:
            return
        visited.add(node)
        stack.add(node)
        for dep in graph.get(node, []):
            dfs(dep, path + [node])
        stack.discard(node)

    for node in graph:
        dfs(node, [])
    return cycles


app_dir = Path("app")
if not app_dir.exists():
    print("✓ Circular deps: no app/ directory")
    sys.exit(0)

graph = build_import_graph(app_dir)
cycles = detect_cycles(graph)

if cycles:
    print(f"✗ Circular dependencies detected ({len(cycles)}):", file=sys.stderr)
    for cycle in cycles:
        names = [Path(f).name for f in cycle]
        print("  " + " → ".join(names), file=sys.stderr)
    sys.exit(1)

print(f"✓ Circular deps: no cycles found in {len(graph)} modules")

File: scripts/governance/check-circular-deps.js
#!/usr/bin/env node
/**
 * Dependency Health — detect circular import cycles in TypeScript source.
 * Uses simple DFS on static import graph.
 */

const fs = require('fs');
const path = require('path');

function buildImportGraph(dir) {
  const graph = {};
  function walk(d) {
    for (const f of fs.readdirSync(d)) {
      const full = path.join(d, f);
      const stat = fs.statSync(full);
      if (stat.isDirectory() && !['node_modules', 'dist', '.git', 'coverage'].includes(f)) {
        walk(full);
      } else if (/\.(ts|tsx)$/.test(f) && !f.endsWith('.d.ts')) {
        const content = fs.readFileSync(full, 'utf8');
        const imports = [];
        const re = /from\s+['"]([^'"]+)['"]/g;
        let m;
        while ((m = re.exec(content)) !== null) {
          const imp = m[1];
          if (imp.startsWith('.')) {
            const resolved = path.resolve(path.dirname(full), imp);
            const candidates = [resolved, resolved + '.ts', resolved + '/index.ts'];
            for (const c of candidates) {
              if (fs.existsSync(c)) {
                imports.push(c);
                break;
              }
            }
          }
        }
        graph[full] = imports;
      }
    }
  }
  walk(dir);
  return graph;
}

function detectCycles(graph) {
  const cycles = [];
  const visited = new Set();
  const stack = new Set();

  function dfs(node, path) {
    if (stack.has(node)) {
      const cycleStart = path.indexOf(node);
      cycles.push(path.slice(cycleStart).concat(node).map(f => path.relative(process.cwd(), f)));
      return;
    }
    if (visited.has(node)) return;
    visited.add(node);
    stack.add(node);
    for (const dep of (graph[node] || [])) {
      dfs(dep, [...path, node]);
    }
    stack.delete(node);
  }

  for (const node of Object.keys(graph)) {
    dfs(node, []);
  }
  return cycles;
}

if (!fs.existsSync('src')) {
  console.log('✓ Circular deps: no src/ directory');
  process.exit(0);
}

const graph = buildImportGraph('src');
const cycles = detectCycles(graph);

if (cycles.length > 0) {
  console.error(`✗ Circular dependencies detected (${cycles.length}):`);
  cycles.forEach(cycle => console.error('  ' + cycle.join(' → ')));
  process.exit(1);
}

console.log(`✓ Circular deps: no cycles found in ${Object.keys(graph).length} modules`);
process.exit(0);
